### DISCUSSION TITLE
GetTrianglesStart/Next strange behaviour.

### DISCUSSION THREAD

**Discussion by itbrandonsilva** [jrouwe/JoltPhysics#836] [Q&A] [answered]
I have the following code that seems to extract the triangles from the capsule shape without issue.

```c++
JPH::Shape* shape = new JPH::CapsuleShape(height, radius);

JPH::Shape::GetTrianglesContext ioContext;
JPH::AABox lb(JPH::Vec3(0, 0, 0), 1000000000);
// This seems to work, as well
//JPH::AABox lb = shape->GetLocalBounds();
shape->GetTrianglesStart(ioContext, lb, JPH::Vec3::sZero(), JPH::Quat::sZero(), JPH::Vec3(1.0f, 1.0f, 1.0f));

JPH::Float3 vertices[10000 * 3];
int count = 0;
while ((count = shape->GetTrianglesNext(ioContext, 10000, vertices)) > 0) {
    // ...
}
``` 

But if I adjust the above with the following, the first call to `GetTrianglesNext` returns 0, and I'm struggling to find the cause.
```c++
JPH::Shape* capsuleShape = new JPH::CapsuleShape(height, radius);
JPH::Shape* shape = new JPH::RotatedTranslatedShape(JPH::Vec3(0, 0, 0), JPH::Quat::sIdentity(), capsuleShape);
```

Looking for a nudge in the right direction.

**Comment by jrouwe** [comment, marked as answer]
If I run your code I get an assert from ```GetTrianglesStart```:

```
Cannot call on non-leaf shapes, use CollectTransformedShapes to collect the leaves first!
```

Does that count as a nudge?

Also:

```
shape->GetTrianglesStart(ioContext, lb, JPH::Vec3::sZero(), JPH::Quat::sZero(), JPH::Vec3(1.0f, 1.0f, 1.0f));
```

Should use ```JPH::Quat::sIdentity()``` instead of ```JPH::Quat::sZero()``` as a zero-ed quaternion doesn't represent a rotation.

**Comment by itbrandonsilva** [comment]
I was initially having difficulty getting my own project to build with a debug version of jolt and have largely been able to make steady progress with using just the release version of jolt. At this point I'm flabbergasted I've made it as far as I have without these assertions. Given what you've shared, I've spent the time necessary to configure my project appropriately to continue working with a debug build of jolt, in which revealed a bunch of other trivial mistakes I've made elsewhere. There was an assertion for the erroneous quaternion you've highlighted, as well. Having resolved all of the assertions, I was eventually able to achieve what I needed.

Now I'm curious about whether or not my approach below is proper for what I'm trying to do. Ultimately when looking to extract the triangles from the `RotatedTranslatedShape`, I was initially expecting the resulting vertices to already be offset based on the transformation info supplied when constructing the shape. Instead, it seems I need to manually supply the translation from the `RotatedTranslatedShape` into `leafShape.GetTrianglesStart`.

```c++
// The way I'm using `CollectTransformedShapes` here is whimsical, as I've yet to spend the time clarifying
// the arguments for myself. I'm aware there is a good chance there is some more nonsense, here.
JPH::SubShapeIDCreator creator;
JPH::AllHitCollisionCollector<JPH::TransformedShapeCollector> collector;
JPH::ShapeFilter filter;
shape->CollectTransformedShapes(lb, JPH::Vec3::sZero(), JPH::Quat::sIdentity(), JPH::Vec3(1.0f, 1.0f, 1.0f), creator, collector, filter); 

for (int i = 0; i < collector.mHits.size(); ++i) {
        JPH::TransformedShape leafShape = collector.mHits[i];
        JPH::Shape::GetTrianglesContext ioContext;

        JPH::Vec3 translation = shape->GetPosition();
        leafShape.GetTrianglesStart(ioContext, lb, -translation);
``` 

Is this the ideal approach for extracting transformed vertices, or is there an alternative?

Appreciate your help! I've otherwise had an enjoyable experience thus far with Jolt despite my ignorance around these parts.

**Reply by jrouwe** [reply]
I think the problem is that CollectTransformedShape takes the COM position, not the world position of the shape so you probably want to do:

```
shape->CollectTransformedShapes(lb, shape->GetCenterOfMass(), ...)
```

(pay attention to the parameter names, these are postfixed by 'COM')

Then you can do:

```
leafShape.GetTrianglesStart(ioContext, lb, RVec3::sZero());
```

**Reply by itbrandonsilva** [reply]
Amazing, that feels much better than what I had. I'll next need to dig and really understand it all...

Thank you so much for your help!
